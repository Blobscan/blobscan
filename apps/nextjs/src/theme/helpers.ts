import type { SystemProps } from "@chakra-ui/styled-system";

/**
 * Styles to disable the native focus ring on chakra components.
 * See https://github.com/chakra-ui/chakra-ui/issues/708
 */
export function chakraFocusRingDisabledStyles(): SystemProps {
  return {
    _focusVisible: { boxShadow: "none", outline: "none" },
  };
}
