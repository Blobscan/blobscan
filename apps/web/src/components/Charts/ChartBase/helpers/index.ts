export * from "./formatters";
export * from "./toolbox";
export * from "./tooltip";
export * from "../types";
