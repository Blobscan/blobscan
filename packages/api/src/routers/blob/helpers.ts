import type {
  BlobscanPrismaClient,
  EthUsdPrice,
  ExtendedBlobDataStorageReferenceSelect,
  Prisma,
} from "@blobscan/db";
import { EthUsdPriceModel } from "@blobscan/db/prisma/zod";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedBlock,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import {
  normalizePrismaBlobDataStorageReferencesFields,
  normalizePrismaTransactionFields,
} from "../../utils";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
  prismaBlobOnTransactionSchema,
} from "../../zod-schemas";

const blobDataStorageReferenceSelect = {
  blobStorage: true,
  url: true,
} satisfies ExtendedBlobDataStorageReferenceSelect;

export const extendedBlobDataStorageReferenceArgs: Prisma.Args<
  BlobscanPrismaClient["blobDataStorageReference"],
  "findFirstOrThrow"
> = {
  select: blobDataStorageReferenceSelect,
  orderBy: {
    blobStorage: "asc",
  },
};

export const baseBlobSelect = {
  commitment: true,
  proof: true,
  size: true,
  usageSize: true,
  versionedHash: true,
} satisfies Prisma.BlobSelect;

export const baseBlobOnTransactionSelect = {
  blobHash: true,
  blockHash: true,
  blockNumber: true,
  blockTimestamp: true,
  index: true,
  txHash: true,
  txIndex: true,
} satisfies Prisma.BlobsOnTransactionsSelect;

export type ExtendedBlobDataStorageReference = NonNullable<
  Prisma.Result<
    BlobscanPrismaClient["blobDataStorageReference"],
    { select: typeof blobDataStorageReferenceSelect },
    "findFirst"
  >
>;
type PrismaBlob = Prisma.BlobGetPayload<{
  select: typeof baseBlobSelect;
}> & {
  dataStorageReferences: ExtendedBlobDataStorageReference[];
};

type PrismaBlobOnTransaction = Prisma.BlobsOnTransactionsGetPayload<{
  select: typeof baseBlobOnTransactionSelect;
}>;

export type CompletedPrismaBlob = Prettify<
  PrismaBlob & {
    transactions: Prettify<
      PrismaBlobOnTransaction & {
        block?: ExpandedBlock;
        transaction?: ExpandedTransaction & {
          block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
        };
      }
    >[];
  }
>;

export type CompletedPrismaBlobOnTransaction = Prettify<
  PrismaBlobOnTransaction & {
    blob: {
      dataStorageReferences: ExtendedBlobDataStorageReference[];
    } & ExpandedBlob;
    block?: ExpandedBlock;
    transaction?: ExpandedTransaction & {
      block: { blobGasPrice: ExpandedBlock["blobGasPrice"] };
    };
  }
>;
export function createBlobSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobSelect,
    dataStorageReferences: extendedBlobDataStorageReferenceArgs,
    transactions: {
      select: {
        ...baseBlobOnTransactionSelect,
        ...blockExpand,
        ...txExpand,
      },
      orderBy: {
        blockNumber: "desc",
      },
    },
  } satisfies Prisma.BlobSelect;
}

export function createBlobsOnTransactionsSelect(expands: Expands) {
  const blockExpand = expands.block ? { block: expands.block } : {};
  const txExpand = expands.transaction
    ? {
        transaction: {
          select: {
            ...expands.transaction.select,
            block: {
              select: {
                blobGasPrice: true,
              },
            },
          },
        },
      }
    : {};

  return {
    ...baseBlobOnTransactionSelect,
    blob: {
      select: {
        ...baseBlobSelect,
        dataStorageReferences: extendedBlobDataStorageReferenceArgs,
      },
    },
    ...txExpand,
    ...blockExpand,
  } satisfies Prisma.BlobsOnTransactionsSelect;
}

const responseBlobOnTxSchema = prismaBlobOnTransactionSchema
  .omit({ blobHash: true })
  .required({
    blockHash: true,
    blockNumber: true,
    blockTimestamp: true,
    index: true,
    txHash: true,
    txIndex: true,
  });

export const responseBlobSchema = baseBlobSchema.extend({
  transactions: z.array(
    responseBlobOnTxSchema
      .merge(
        baseTransactionSchema
          .omit({
            hash: true,
            blockNumber: true,
            blockTimestamp: true,
            blockHash: true,
          })
          .partial()
      )
      .extend({
        block: baseBlockSchema
          .omit({
            hash: true,
            number: true,
            timestamp: true,
          })
          .partial({
            blobGasBaseUsdFee: true,
            blobGasUsdPrice: true,
          })
          .optional(),
        ethUsdPrice: EthUsdPriceModel.shape.price.optional(),
      })
  ),
});

export const responseBlobOnTransactionSchema = responseBlobOnTxSchema
  .merge(baseBlobSchema)
  .extend({
    block: baseBlockSchema
      .omit({
        hash: true,
        number: true,
        timestamp: true,
      })
      .optional(),
    transaction: baseTransactionSchema
      .omit({
        hash: true,
        blockNumber: true,
        blockTimestamp: true,
        blockHash: true,
      })
      .optional(),
  });

export type ResponseBlob = z.input<typeof responseBlobSchema>;
export type ResponseBlobOnTransaction = z.input<
  typeof responseBlobOnTransactionSchema
>;

export function toResponseBlob(
  {
    dataStorageReferences,
    transactions: blobOnTxs,
    ...prismaBlob
  }: CompletedPrismaBlob,
  ethUsdPrices: EthUsdPrice["price"][]
): ResponseBlob {
  const responseBlob: ResponseBlob = {
    ...prismaBlob,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    transactions: blobOnTxs.map(
      (
        {
          block: { computeUsdFields, ...restBlock } = {},
          transaction: prismaTx = {},
          ...restBlobOnTx
        },
        i
      ) => {
        const txEthUsdPrice = ethUsdPrices[i];

        const blockUsdFields =
          txEthUsdPrice && computeUsdFields
            ? computeUsdFields(txEthUsdPrice)
            : undefined;
        const txBlock = {
          ...restBlock,
          ...(blockUsdFields ?? {}),
        };

        let responseBlobOnTx: ResponseBlob["transactions"][number] = {
          ...restBlobOnTx,
          ...txBlock,
          ethUsdPrice: txEthUsdPrice?.toNumber(),
        };

        if (isFullyDefined(prismaTx)) {
          const {
            block,
            decodedFields,
            from,
            fromId,
            toId,
            computeBlobGasBaseFee,
            computeUsdFields,
            ...restPrismaTx
          } = prismaTx;

          responseBlobOnTx = {
            ...responseBlobOnTx,
            ...restPrismaTx,
            ...normalizePrismaTransactionFields({
              decodedFields,
              from,
              fromId,
              toId,
            }),
            blobGasBaseFee: computeBlobGasBaseFee(block.blobGasPrice),
            ...(txEthUsdPrice
              ? computeUsdFields({
                  blobGasPrice: block.blobGasPrice,
                  ethUsdPrice: txEthUsdPrice,
                })
              : {}),
          };
        }

        return responseBlobOnTx;
      }
    ),
  };

  return responseBlob;
}

export function toResponseBlobOnTransaction(
  prismaBlobOnTransaction: CompletedPrismaBlobOnTransaction,
  ethUsdPrice?: EthUsdPrice["price"]
): ResponseBlobOnTransaction {
  const {
    blobHash,
    blob: { dataStorageReferences, ...blob },
    transaction,
    block,
    ...restBlobOnTransaction
  } = prismaBlobOnTransaction;

  const responseBlobOnTx: ResponseBlobOnTransaction = {
    ...blob,
    versionedHash: blobHash,
    dataStorageReferences: normalizePrismaBlobDataStorageReferencesFields(
      dataStorageReferences
    ),
    ...restBlobOnTransaction,
  };

  if (transaction) {
    const {
      block: txBlock,
      decodedFields,
      from,
      fromId,
      toId,
      computeBlobGasBaseFee,
      computeUsdFields,
      ...restTx
    } = transaction;

    responseBlobOnTx.transaction = {
      ...restTx,
      blobGasPrice: txBlock.blobGasPrice,
      ...normalizePrismaTransactionFields({
        decodedFields,
        from,
        fromId,
        toId,
      }),
      blobGasBaseFee: computeBlobGasBaseFee(txBlock.blobGasPrice),
      ...(ethUsdPrice
        ? computeUsdFields({
            blobGasPrice: txBlock.blobGasPrice,
            ethUsdPrice,
          })
        : {}),
    };
  }

  if (block) {
    const { computeUsdFields, ...restBlock } = block;
    responseBlobOnTx.block = {
      ...restBlock,
      ...(ethUsdPrice ? computeUsdFields(ethUsdPrice) : {}),
    };
  }

  return responseBlobOnTx;
}
