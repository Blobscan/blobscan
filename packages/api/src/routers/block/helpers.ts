import type { BlobStorageManager } from "@blobscan/blob-storage-manager";
import type { BlobscanPrismaClient, Prisma } from "@blobscan/db";
import { z } from "@blobscan/zod";

import type {
  ExpandedBlob,
  ExpandedTransaction,
  Expands,
} from "../../middlewares/withExpands";
import type { Filters } from "../../middlewares/withFilters";
import type { Prettify } from "../../types";
import { isFullyDefined } from "../../utils";
import { retrieveBlobData } from "../../utils";
import {
  deriveTransactionFields,
  normalizePrismaBlobFields,
  normalizePrismaTransactionFields,
} from "../../utils/transformers";
import {
  baseBlobSchema,
  baseBlockSchema,
  baseTransactionSchema,
} from "../../zod-schemas";

const prismaBlockSelect = {
  hash: true,
  number: true,
  timestamp: true,
  slot: true,
  blobGasUsed: true,
  blobAsCalldataGasUsed: true,
  blobGasPrice: true,
  excessBlobGas: true,
} satisfies Prisma.BlockSelect;

export type PrismaBlock = Prisma.BlockGetPayload<{
  select: typeof prismaBlockSelect;
}>;

export type CompletePrismaBlock = Prettify<
  PrismaBlock & {
    transactions: Prettify<
      {
        hash: string;
      } & Partial<ExpandedTransaction> & {
          blobs: Prettify<
            { blobHash: string } & {
              blob?: ExpandedBlob;
            }
          >[];
        }
    >[];
  }
>;

export function createBlockSelect(expands: Expands, filters?: Filters) {
  const blobExpand = expands.blob ? { blob: expands.blob } : {};
  const transactionExpand = expands.transaction?.select ?? {};

  return {
    ...prismaBlockSelect,
    transactions: {
      select: {
        hash: true,
        ...transactionExpand,
        blobs: {
          select: {
            blobHash: true,
            ...blobExpand,
          },
          orderBy: {
            index: "asc",
          },
        },
      },
      orderBy: {
        index: "asc",
      },
      where: filters?.transactionFilters,
    },
  } satisfies Prisma.BlockSelect;
}

export const responseBlockSchema = baseBlockSchema.extend({
  transactions: z.array(
    baseTransactionSchema
      .omit({
        blockHash: true,
        blockNumber: true,
        blockTimestamp: true,
      })
      .partial()
      .required({ hash: true })
      .extend({
        blobs: z.array(
          baseBlobSchema.partial().required({ versionedHash: true })
        ),
      })
  ),
});

export type ResponseBlock = z.input<typeof responseBlockSchema>;

export function toResponseBlock(
  prismaBlock: CompletePrismaBlock
): ResponseBlock {
  const transactions = prismaBlock.transactions.map(
    ({ hash, blobs, ...prismaTx }) => ({
      hash,
      ...(isFullyDefined(prismaTx)
        ? {
            ...prismaTx,
            ...normalizePrismaTransactionFields(prismaTx),
            ...deriveTransactionFields({
              ...prismaTx,
              blobGasPrice: prismaBlock.blobGasPrice,
            }),
          }
        : {}),
      blobs: blobs.map(({ blobHash, blob }) =>
        blob
          ? normalizePrismaBlobFields({ versionedHash: blobHash, ...blob })
          : { versionedHash: blobHash }
      ),
    })
  );

  return {
    ...prismaBlock,
    transactions,
  };
}

export type BlockId = "hash" | "number" | "slot";
export type BlockIdField =
  | { type: "hash"; value: string }
  | { type: "number"; value: number }
  | { type: "slot"; value: number };

function buildBlockWhereClause(
  { type, value }: BlockIdField,
  filters: Filters
): Prisma.BlockWhereInput {
  switch (type) {
    case "hash": {
      return { hash: value };
    }
    case "number": {
      return { number: value, transactionForks: filters.blockType };
    }
    case "slot": {
      return { slot: value, transactionForks: filters.blockType };
    }
  }
}

export async function fetchBlock(
  blockId: BlockIdField,
  {
    blobStorageManager,
    prisma,
    filters,
    expands,
  }: {
    blobStorageManager: BlobStorageManager;
    prisma: BlobscanPrismaClient;
    filters: Filters;
    expands: Expands;
  }
): Promise<CompletePrismaBlock | undefined> {
  const select = createBlockSelect(expands);
  const where = buildBlockWhereClause(blockId, filters);

  const prismaBlock = (await prisma.block.findFirst({
    select,
    where,
  })) as unknown as CompletePrismaBlock | null;

  if (!prismaBlock) {
    return;
  }

  if (expands.blobData) {
    await Promise.all(
      prismaBlock.transactions
        .flatMap((tx) => tx.blobs)
        .map(async ({ blobHash, blob }) => {
          const dataStorageReferences = blob?.dataStorageReferences;
          if (dataStorageReferences) {
            const data = await retrieveBlobData(blobStorageManager, {
              dataStorageReferences,
              versionedHash: blobHash,
            });

            blob.data = data;
          }
        })
    );
  }

  return prismaBlock;
}
