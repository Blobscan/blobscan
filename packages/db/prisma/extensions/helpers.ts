import { Prisma } from "@prisma/client";

export type ZeroOpResult = { count: number }[];

export const helpersExtension = Prisma.defineExtension((prisma) =>
  prisma.$extends({
    name: "Helper Operations",
    model: {
      $allModels: {
        zero() {
          return prisma.$queryRaw<ZeroOpResult>`SELECT 0 as count`;
        },
      },
      block: {
        findLatest() {
          return prisma.block.findFirst({
            where: {
              transactionForks: {
                none: {},
              },
            },
            orderBy: { number: "desc" },
          });
        },
      },
    },
  })
);
