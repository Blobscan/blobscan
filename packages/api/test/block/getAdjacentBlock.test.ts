import { TRPCError } from "@trpc/server";
import { beforeAll, describe, expect, it } from "vitest";

import { testValidError } from "@blobscan/test";

import type { TRPCContext } from "../../src";
import { createTestContext, runExpandsTestsSuite } from "../helpers";
import type { BlockCaller } from "./caller";
import { createBlockCaller } from "./caller";

describe("getAdjacentBlock", () => {
  let ctx: TRPCContext;
  let caller: BlockCaller;
  const blockNumber = 1004;

  beforeAll(async () => {
    ctx = await createTestContext();
    caller = createBlockCaller(ctx);
  });

  runExpandsTestsSuite("block", ["transaction", "blob"], (expandInput) =>
    caller.getAdjacentBlock({
      blockNumber: 1004,
      direction: "next",
      ...expandInput,
    })
  );

  it("should get a previous block", async () => {
    const previousBlock = await caller.getAdjacentBlock({
      blockNumber,
      direction: "prev",
    });

    expect(previousBlock).toMatchSnapshot();
  });

  it("should get a next block", async () => {
    const nextBlock = await caller.getAdjacentBlock({
      blockNumber,
      direction: "next",
    });

    expect(nextBlock).toMatchSnapshot();
  });

  it("should ignore reorged blocks", async () => {
    const nextBlock = await caller.getAdjacentBlock({
      blockNumber: 1007,
      direction: "next",
    });

    expect(nextBlock.hash).toBe(
      "0x8000000000000000000000000000000000000000000000000000000000000000"
    );
  });

  testValidError(
    "should fail when no adjacent block was found",
    async () => {
      await caller.getAdjacentBlock({
        blockNumber: 1001,
        direction: "prev",
      });
    },
    TRPCError,
    {
      checkCause: true,
    }
  );
});
