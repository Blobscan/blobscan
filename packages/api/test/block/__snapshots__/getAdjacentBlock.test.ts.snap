// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getAdjacentBlock > should fail when no adjacent block was found 1`] = `"No prev block found for block 1001"`;

exports[`getAdjacentBlock > should fail when no adjacent block was found 2`] = `undefined`;

exports[`getAdjacentBlock > should get a next block 1`] = `
{
  "blobAsCalldataGasUsed": 250000n,
  "blobGasBaseFee": 100000000n,
  "blobGasPrice": 20n,
  "blobGasUsed": 5000000n,
  "ethUsdPrice": undefined,
  "excessBlobGas": 10000n,
  "hash": "blockHash005",
  "number": 1005,
  "slot": 105,
  "timestamp": 2023-08-28T10:00:00.000Z,
  "transactions": [
    {
      "blobs": [
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "versionedHash": "blobHash001",
        },
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "versionedHash": "blobHash001",
        },
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "versionedHash": "blobHash001",
        },
      ],
      "hash": "txHash012",
    },
    {
      "blobs": [
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "versionedHash": "blobHash001",
        },
      ],
      "hash": "txHash013",
    },
  ],
}
`;

exports[`getAdjacentBlock > should get a previous block 1`] = `
{
  "blobAsCalldataGasUsed": 255000n,
  "blobGasBaseFee": 121000000n,
  "blobGasPrice": 22n,
  "blobGasUsed": 5500000n,
  "ethUsdPrice": undefined,
  "excessBlobGas": 15000n,
  "hash": "blockHash003",
  "number": 1003,
  "slot": 103,
  "timestamp": 2023-08-03T12:00:00.000Z,
  "transactions": [
    {
      "blobs": [
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash002.bin?alt=media",
            },
          ],
          "versionedHash": "blobHash002",
        },
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash002.bin?alt=media",
            },
          ],
          "versionedHash": "blobHash002",
        },
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "versionedHash": "blobHash001",
        },
      ],
      "hash": "txHash005",
    },
    {
      "blobs": [
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "versionedHash": "blobHash001",
        },
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "versionedHash": "blobHash001",
        },
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "versionedHash": "blobHash001",
        },
      ],
      "hash": "txHash006",
    },
  ],
}
`;

exports[`getAdjacentBlock > when getting expanded block results > should return the correct expanded blob 1`] = `
{
  "blobAsCalldataGasUsed": 250000n,
  "blobGasBaseFee": 100000000n,
  "blobGasPrice": 20n,
  "blobGasUsed": 5000000n,
  "ethUsdPrice": undefined,
  "excessBlobGas": 10000n,
  "hash": "blockHash005",
  "number": 1005,
  "slot": 105,
  "timestamp": 2023-08-28T10:00:00.000Z,
  "transactions": [
    {
      "blobs": [
        {
          "commitment": "commitment001",
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "proof": "proof001",
          "size": 1000,
          "usageSize": 1000,
          "versionedHash": "blobHash001",
        },
        {
          "commitment": "commitment001",
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "proof": "proof001",
          "size": 1000,
          "usageSize": 1000,
          "versionedHash": "blobHash001",
        },
        {
          "commitment": "commitment001",
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "proof": "proof001",
          "size": 1000,
          "usageSize": 1000,
          "versionedHash": "blobHash001",
        },
      ],
      "hash": "txHash012",
    },
    {
      "blobs": [
        {
          "commitment": "commitment001",
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "proof": "proof001",
          "size": 1000,
          "usageSize": 1000,
          "versionedHash": "blobHash001",
        },
      ],
      "hash": "txHash013",
    },
  ],
}
`;

exports[`getAdjacentBlock > when getting expanded block results > should return the correct expanded results when multiple expands are requested 1`] = `
{
  "blobAsCalldataGasUsed": 250000n,
  "blobGasBaseFee": 100000000n,
  "blobGasPrice": 20n,
  "blobGasUsed": 5000000n,
  "ethUsdPrice": undefined,
  "excessBlobGas": 10000n,
  "hash": "blockHash005",
  "number": 1005,
  "slot": 105,
  "timestamp": 2023-08-28T10:00:00.000Z,
  "transactions": [
    {
      "blobAsCalldataGasFee": 10000n,
      "blobAsCalldataGasUsed": 1000n,
      "blobGasBaseFee": 7864320n,
      "blobGasMaxFee": 39321600n,
      "blobGasUsed": 393216n,
      "blobs": [
        {
          "commitment": "commitment001",
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "proof": "proof001",
          "size": 1000,
          "usageSize": 1000,
          "versionedHash": "blobHash001",
        },
        {
          "commitment": "commitment001",
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "proof": "proof001",
          "size": 1000,
          "usageSize": 1000,
          "versionedHash": "blobHash001",
        },
        {
          "commitment": "commitment001",
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "proof": "proof001",
          "size": 1000,
          "usageSize": 1000,
          "versionedHash": "blobHash001",
        },
      ],
      "category": "rollup",
      "decodedFields": null,
      "from": "0x6887246668a3b87f54deb3b94ba47a6f63f32985",
      "hash": "txHash012",
      "index": 0,
      "maxFeePerBlobGas": 100n,
      "rollup": "optimism",
      "to": "address4",
    },
    {
      "blobAsCalldataGasFee": 10000n,
      "blobAsCalldataGasUsed": 1000n,
      "blobGasBaseFee": 2621440n,
      "blobGasMaxFee": 13107200n,
      "blobGasUsed": 131072n,
      "blobs": [
        {
          "commitment": "commitment001",
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "proof": "proof001",
          "size": 1000,
          "usageSize": 1000,
          "versionedHash": "blobHash001",
        },
      ],
      "category": "other",
      "decodedFields": null,
      "from": "address5",
      "hash": "txHash013",
      "index": 1,
      "maxFeePerBlobGas": 100n,
      "rollup": null,
      "to": "address4",
    },
  ],
}
`;

exports[`getAdjacentBlock > when getting expanded block results > should return the correct expanded transaction 1`] = `
{
  "blobAsCalldataGasUsed": 250000n,
  "blobGasBaseFee": 100000000n,
  "blobGasPrice": 20n,
  "blobGasUsed": 5000000n,
  "ethUsdPrice": undefined,
  "excessBlobGas": 10000n,
  "hash": "blockHash005",
  "number": 1005,
  "slot": 105,
  "timestamp": 2023-08-28T10:00:00.000Z,
  "transactions": [
    {
      "blobAsCalldataGasFee": 10000n,
      "blobAsCalldataGasUsed": 1000n,
      "blobGasBaseFee": 7864320n,
      "blobGasMaxFee": 39321600n,
      "blobGasUsed": 393216n,
      "blobs": [
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "versionedHash": "blobHash001",
        },
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "versionedHash": "blobHash001",
        },
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "versionedHash": "blobHash001",
        },
      ],
      "category": "rollup",
      "decodedFields": null,
      "from": "0x6887246668a3b87f54deb3b94ba47a6f63f32985",
      "hash": "txHash012",
      "index": 0,
      "maxFeePerBlobGas": 100n,
      "rollup": "optimism",
      "to": "address4",
    },
    {
      "blobAsCalldataGasFee": 10000n,
      "blobAsCalldataGasUsed": 1000n,
      "blobGasBaseFee": 2621440n,
      "blobGasMaxFee": 13107200n,
      "blobGasUsed": 131072n,
      "blobs": [
        {
          "dataStorageReferences": [
            {
              "storage": "google",
              "url": "http://localhost:4443/storage/v1/b/blobscan-test-bucket/o/1%2Fob%2FHa%2Fsh%2FobHash001.bin?alt=media",
            },
            {
              "storage": "swarm",
              "url": "https://api.gateway.ethswarm.org/bzz/bzz://some-hash-for-blobHash001",
            },
          ],
          "versionedHash": "blobHash001",
        },
      ],
      "category": "other",
      "decodedFields": null,
      "from": "address5",
      "hash": "txHash013",
      "index": 1,
      "maxFeePerBlobGas": 100n,
      "rollup": null,
      "to": "address4",
    },
  ],
}
`;
